### PR TITLE
#37 - Support for Spock

### PR DESCRIPTION
**Pull request by phejl**
Hopefully this could do the job.
- `awaitEvents()` is called on stream access. Not sure if there is a better way, but it seems that's what one typically do anyway even in JUnit. Perhaps there is a room for optimization.
- `JfrEvents` instance is cleared after each test unless marked as `@Shared`.

### CODE REVIEW

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (leonard84) [inline src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java:34]: Depending on how you want to deal with inheritance, you might want to call `spec.getBottomSpec().getAllFeatures()` instead.

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (leonard84) [inline src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java:59]: Why is this not a static class/in its own file?

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (leonard84) [inline src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java:57]: These methods belong to the interceptor

File: README.md
Comment (leonard84) [inline README.md:90]: Adding an actual test would be a good idea, and I'd also suggest to test a data-driven feature in combination with this extension.

**Review by leonard84 — CHANGES_REQUESTED** [review]

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java]: Thanks for your suggestion!

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java]: Yep that would be better. Changed.

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java]: Done.

File: README.md
Comment (phejl) [reply README.md:90]: I'll try to add some test later tonight/tomorrow.

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (leonard84) [inline src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java:56]: Spock creates a fresh instance of the test class and thus its fields for each run, so calling `clear()` is probably unnecessary.

**Comment by gunnarmorling** [comment]
Thanks a lot for this! Could we add an actual test which makes use of this, too? Maybe something under _src/it/groovy_ or so?

**Comment by phejl** [comment]
> Thanks a lot for this! Could we add an actual test which makes use of this, too? Maybe something under _src/it/groovy_ or so?

Yep, working on it.

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java]: Thanks. Removed and code simplified.

File: README.md
Comment (phejl) [reply README.md:90]: Leonard, @gunnarmorling I've added more tests.

**Comment by phejl** [comment]
I hopefully resolved all the comments so far.

File: src/test/groovy/dev/morling/jfrunit/JfrSpockSharedSpec.groovy
Comment (leonard84) [inline src/test/groovy/dev/morling/jfrunit/JfrSpockSharedSpec.groovy:38]:
```suggestion
    def 'records written bytes on each iteration'(int iteration) {
```

I prefer imperative style

File: src/test/groovy/dev/morling/jfrunit/JfrSpockSharedSpec.groovy
Comment (leonard84) [inline src/test/groovy/dev/morling/jfrunit/JfrSpockSharedSpec.groovy:49]:
```suggestion
        jfrEvents.events.filter{ it.eventType.name == 'jdk.FileWrite' }.count() == iteration
```

File: src/test/groovy/dev/morling/jfrunit/JfrSpockSpec.groovy
Comment (leonard84) [inline src/test/groovy/dev/morling/jfrunit/JfrSpockSpec.groovy:40]: For tests, I would try to use shorter sleeps if possible.

Comment (phejl) [reply src/test/groovy/dev/morling/jfrunit/JfrSpockSpec.groovy:40]: This one is matching the tests used for JUnit. Should I change even there?

File: pom.xml
Comment (phejl) [reply pom.xml:171]: I know but I was used to have `Spec` for Spock tests. I can change it if you insist on that.

File: pom.xml
Comment (phejl) [reply pom.xml]: Done.

File: src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEventTestSpockExtension.java]: Done.

File: src/test/groovy/dev/morling/jfrunit/JfrSpockSpec.groovy
Comment (phejl) [reply src/test/groovy/dev/morling/jfrunit/JfrSpockSpec.groovy]: I've lowered sleeps across the tests.

File: pom.xml
Comment (leonard84) [reply pom.xml:170]: That is up to @gunnarmorling

**Comment by phejl** [comment]
@leonard84 @gunnarmorling Let me know if I should do anything more. I'm used to leaving up comment resolution to the reporter. Please, let me know if you prefer it in a different way. Thanks in advance.

**Comment by leonard84** [comment]
I've given my feedback, but as I'm not a committer of this project @gunnarmorling should do a full review anyway. I haven't seen new issues, so you can go ahead and mark all uncontested comments as resolved.

**Comment by phejl** [comment]
Hi @gunnarmorling I'm wondering whether there is anything I should do at the moment. I think I would have some follow up PRs later :).

**Comment by phejl** [comment]
Hello @gunnarmorling , I don't want to be pushy. Just please let me know whether there is something I should do or whether it is just - completely understandable - lack of time.

File: src/main/java/dev/morling/jfrunit/JfrEvents.java
Comment (gunnarmorling) [inline src/main/java/dev/morling/jfrunit/JfrEvents.java:135]: What's the motivation for exposing this in addition to `getEvents()`?

**Review by gunnarmorling — APPROVED** [review]
Hey @phejl, first of all, apologies for the long delay on my end, there's just too many things going on :(

The change looks good to me, thanks a lot for this awesome contribution. Two questions/comments inline; also, how does sychronization between running the test code and the assertions, without the explicit `awaitEvents()` barrier?

Slightly related, I remember you saying you have one more change lined up. Could you log a separate GH issue for this one? If you are interested in contributing continously, I'll also happily make you a committer to this repo. Just let me know. Thanks again!

File: src/main/java/dev/morling/jfrunit/JfrEvents.java
Comment (phejl) [reply src/main/java/dev/morling/jfrunit/JfrEvents.java:135]: For the `*Spec` I've always seen it like that with Spock. Sometimes people use `*Specification`. I've never saw `*Test`. Also Spock's methods suggest a spec by having `setupSpec()` and `cleanupSpec()`. But I can live with anything you would like :)

> What's the motivation for exposing this in addition to getEvents()?

**Comment by phejl** [comment]
> Slightly related, I remember you saying you have one more change lined up. Could you log a separate GH issue for this one? If you are interested in contributing continously, I'll also happily make you a committer to this repo. Just let me know. Thanks again!

@gunnarmorling Yes, I'll do. I just wanted to wait for this in case it would awfully wrong. Beyond the next one I don't have anything else in the queue so leaving the decision up to you.

...and thank you.

File: src/main/java/dev/morling/jfrunit/JfrEvents.java
Comment (gunnarmorling) [reply src/main/java/dev/morling/jfrunit/JfrEvents.java:135]: > For the `*Spec` I've always seen it like that with Spock.

Ok, cool, let's follow this pattern then.

> I don't think I'm exposing anything new.

Hum, yes, you're right, it was like that before. Which raises the question: should we make `stream()` private, and perhaps rename `getEvents()` into `events()` (or `recordedEvents()`), for more consistency and exposing only a single public method for the purpose of getting the events?

**Comment by gunnarmorling** [comment]
@phejl, so I think we're good to go, apart from the last suggestion above about renaming and making `stream()` private. And this will need a rebase + force push due to the conflict in _pom.xml_.

> Beyond the next one I don't have anything else in the queue so leaving the decision up to you.

I'll add you as a committer just in case; the more, the merrier :) Thanks a lot for your efforts!

**Comment by phejl** [comment]
> @phejl, so I think we're good to go, apart from the last suggestion above about renaming and making `stream()` private. And this will need a rebase + force push due to the conflict in _pom.xml_.

I was going to do that last evening but I'm suffering from test failures (related to ThreadSleep events and multiple calls to `awaitEvents`). I'll need some more time to figure out what's going on.

**Comment by gunnarmorling** [comment]
If you haven't done so yet, I'd suggest you rebase this to the current master first and see whether that helps. I was seeing some test failure too which I was addressing while moving to Java 16 as a baseline.

**Comment by phejl** [comment]
@gunnarmorling I should have written that - it was after the rebase. Anyway I think I have found the cause. I believe we need to set `stream.setReuse(false)` otherwise JFR might reuse the event we captured and we hold the reference to. This might lead to strange behaviour that captured event (such as jdk.ThreadSleep) might change under the hood while still being the same instance.

Beyond that I changed the internal wait so it is hopefully less probable to match real world apps waits. And of course I made `stream()` private as we agreed before.

I hope it should be good to go finally.

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
And merged. Thanks a ton, @phejl 🎉 !

**Comment by phejl** [comment]
Thank you @gunnarmorling!

**Pull request merged**
